frame/referenda/src/types.rs

```diff
@@ -282,21 +249,169 @@ impl<
 #[derive(Clone, Eq, PartialEq, Encode, Decode, TypeInfo, MaxEncodedLen)]
 #[cfg_attr(not(feature = "std"), derive(RuntimeDebug))]
 pub enum Curve {
-	/// Linear curve starting at `(0, begin)`, ending at `(period, begin - delta)`.
-	LinearDecreasing { begin: Perbill, delta: Perbill },
+	/// Linear curve starting at `(0, ceil)`, proceeding linearly to `(length, floor)`, then
+	/// remaining at `floor` until the end of the period.
+	LinearDecreasing { length: Perbill, floor: Perbill, ceil: Perbill },
+	/// Stepped curve, beginning at `(0, begin)`, then remaining constant for `period`, at which
+	/// point it steps down to `(period, begin - step)`. It then remains constant for another
+	/// `period` before stepping down to `(period * 2, begin - step * 2)`. This pattern continues
+	/// but the `y` component has a lower limit of `end`.
+	SteppedDecreasing { begin: Perbill, end: Perbill, step: Perbill, period: Perbill },
+	/// A recipocal (`K/(x+S)-T`) curve: `factor` is `K` and `x_offset` is `S`, `y_offset` is `T`.
+	Reciprocal { factor: FixedI64, x_offset: FixedI64, y_offset: FixedI64 },
+}
+
+const fn pos_quad_solution(a: FixedI64, b: FixedI64, c: FixedI64) -> FixedI64 {
+	const TWO: FixedI64 = FixedI64::from_u32(2);
+	const FOUR: FixedI64 = FixedI64::from_u32(4);
+	b.neg().add(b.mul(b).sub(FOUR.mul(a).mul(c)).sqrt()).div(TWO.mul(a))
```

Comment:
Do these have saturated or checked equivalents? Feels like this could easily panic if e.g. `a = 0`.

Reply:
Also, since this is only calculating 1 root, how does the function calculate a concave up/down curve for a reciprocal? I'm assuming 1 root only calculates a concave up curve, and that we need the 2nd root to calculate a concave down.

Reply:
I'll add some docs. As a `const` function it is meant for use at compile-time and panicking is acceptable.